### PR TITLE
Allow sending empty collection by skipping the actual send

### LINKED ISSUE #550: ISenderClient:SendAsync(IList<Message> messageList) throws null reference when messageList is empty
## Actual Behavior
1. SendAsync throws null reference exception when given collection of messages is empty.

## Expected Behavior
1. No exception should be thrown.

## Simplest repro
```csharp
        [Fact]
        [DisplayTestMethodName]
        void Empty_Collection_Of_Messages_Returns_Empty_AMQP_Message()
        {
            var messages = new Message[0];

            var batchedMessage = AmqpMessageConverter.BatchSBMessagesAsAmqpMessage(messages);

            Assert.NotNull(batchedMessage);
        }
```

> System.NullReferenceException : Object reference not set to an instance of an object.
>    at Microsoft.Azure.ServiceBus.Amqp.AmqpMessageConverter.BatchSBMessagesAsAmqpMessage(IEnumerable`1 sbMessages)
>    at Microsoft.Azure.ServiceBus.UnitTests.AmqpConverterTests.Empty_Collection_Of_Messages_Returns_Empty_AMQP_Message()

## Reason
https://github.com/Azure/azure-service-bus-dotnet/blob/6f144e91310dcc7bd37aba4e8aebd535d13fa31a/src/Microsoft.Azure.ServiceBus/Amqp/AmqpMessageConverter.cs#L59
`dataList` is initalized in a foreach loop, so when there is no messages it will be null.
Later one, its used by `amqpMessage = AmqpMessage.Create(dataList);` what causes null reference exception.

## Versions  
- OS platform and version: Windows 10
- .NET Version: .NET Core 2.1
- NuGet package version or commit ID: 3.1.0-preview, master

I can pick up this issue if you want.

### PR DESCRIPTION
**Pull request by SeanFeldman**
Fixes #550

### CODE REVIEW

File: test/Microsoft.Azure.ServiceBus.UnitTests/SenderReceiverTests.cs [hunk 1]
```diff
@@ -459,5 +459,16 @@ public async Task ClientsUseGlobalConnectionCloseFirstClientSecoundClientShouldS
                 await receiver.CloseAsync().ConfigureAwait(false);
             }
         }
+
+        [Theory]
+        [InlineData(TestConstants.NonPartitionedQueueName)]
+        [DisplayTestMethodName]
+        async Task MessageSenderThrowsWhenSendingEmptyCollection(string queueName)
```
Comment (thomaslevesque) [inline test/Microsoft.Azure.ServiceBus.UnitTests/SenderReceiverTests.cs]: You might want to rename this method to reflect what is actually being tested

Comment (SeanFeldman) [reply test/Microsoft.Azure.ServiceBus.UnitTests/SenderReceiverTests.cs]: WIP. Also, I'm thinking the test is not good enough 🙂

**Comment by SeanFeldman** [comment]
@nemakam please review

File: test/Microsoft.Azure.ServiceBus.UnitTests/SenderReceiverTests.cs (hunk 1 above)
Comment (nemakam) [reply test/Microsoft.Azure.ServiceBus.UnitTests/SenderReceiverTests.cs]: Yeah the test somehow doesnt look perfect. Why do we even need a receiver here? Of course there is no message sent as its an empty list. Its not worth spending those 3 seconds * number of build configurations to test this.

---
In reply to: [254793100](https://github.com/Azure/azure-service-bus-dotnet/pull/642#discussion_r254793100) [](ancestors = 254793100)

File: src/Microsoft.Azure.ServiceBus/Core/MessageSender.cs [hunk 2]
```diff
@@ -453,6 +458,11 @@ static int ValidateMessages(IList<Message> messageList)
                 throw Fx.Exception.ArgumentNull(nameof(messageList));
             }
 
+            if (messageList.Count == 0)
```
Comment (nemakam) [inline src/Microsoft.Azure.ServiceBus/Core/MessageSender.cs]: >if (messageList.Count == 0) [](start = 12, length = 27)

This isnt required

File: test/Microsoft.Azure.ServiceBus.UnitTests/SenderReceiverTests.cs (hunk 1 above)
Comment (SeanFeldman) [reply test/Microsoft.Azure.ServiceBus.UnitTests/SenderReceiverTests.cs]: Any ideas how to test it better? I'm open for suggestions.

File: src/Microsoft.Azure.ServiceBus/Core/MessageSender.cs (hunk 2 above)
Comment (SeanFeldman) [reply src/Microsoft.Azure.ServiceBus/Core/MessageSender.cs]: As in let `foreach` run with zero iterations? Could do that, but then it's less obvious what will happen, no?

**Comment by SeanFeldman** [comment]
Marked PR as `[WIP]` until it's not and ready to be merged.

**Comment by SeanFeldman** [comment]
Marked PR as `[WIP]` until it's not and ready to be merged.

**Comment by dlstucki** [comment]
@nemakam, et. al., what's the explanation for not throwing ArgumentException instead? Is it ease of the implementation of this library method? Most APIs I can think of which have some kind of "Send" method tend to throw if you tell it to send but don't give any data for sending. I feel like silently doing nothing might be hiding errors in the code calling this library.

If it's ease of implementation (like avoiding enumerating something twice) could something like this be done?
```
SendBatch(IEnumerable<Whatever> batch)
{
    bool foundAny = false;
    foreach (Whatever w in batch)
    {
        foundAny = true;
        // ... Do something ...
    }

    if (!foundAny) { throw new ArgumentException("batch", "Batch contained no items");
}
```

**Comment by SeanFeldman** [comment]
@dlstucki that was my [initial take](https://github.com/Azure/azure-service-bus-dotnet/pull/641), but there was a compelling argument why an empty list should not throw: https://github.com/Azure/azure-service-bus-dotnet/issues/550#issuecomment-461330073

**Comment by dlstucki** [comment]
> @dlstucki that was my [initial take]

@SeanFeldman Thanks! Seeing other APIs which behave this way makes me feel better.

**Comment by SeanFeldman** [comment]
The only reason this is still not merged is because I have no idea what would be a better way to test this behaviour w/o hitting the broker. If you have ideas @dlstucki, please share 🙂

**Comment by dlstucki** [comment]
@SeanFeldman, I haven't been following along very closely so I probably don't understand it. What do you need to verify in order to test this?  Would this work?
- Sending 1 message before
- Send an empty batch (make sure it doesn't throw)
- Sending 1 message after
- Verify the before and after messages are received successfully

**Comment by SeanFeldman** [comment]
@dlstucki that's sort of [in place](https://github.com/Azure/azure-service-bus-dotnet/pull/642/files#diff-649e3f0fe912cbbd18f4ba770db10005R466) already.

File: src/Microsoft.Azure.ServiceBus/Core/MessageSender.cs (hunk 2 above)
Comment (SeanFeldman) [reply src/Microsoft.Azure.ServiceBus/Core/MessageSender.cs]: Done

**Comment by SeanFeldman** [comment]
I'll go with this for now and if a better test idea emerges, can be PR-ed

**Pull request merged**